hail/src/main/scala/is/hail/expr/ir/ExtractIntervalFilters.scala

```diff
@@ -2,32 +2,75 @@ package is.hail.expr.ir
 import is.hail.annotations.IntervalEndpointOrdering
+import is.hail.rvd.PartitionBoundOrdering
 import is.hail.types.virtual._
 import is.hail.utils.{FastSeq, Interval, IntervalEndpoint, _}
 import is.hail.variant.{Locus, ReferenceGenome}
 import org.apache.spark.sql.Row
 
+import scala.Option.option2Iterable
+
 object ExtractIntervalFilters {
 
   val MAX_LITERAL_SIZE = 4096
 
-  case class ExtractionState(ctx: ExecuteContext, rowRef: Ref, keyFields: IndexedSeq[String]) {
+  object ExtractionState {
+    def apply(
+      ctx: ExecuteContext,
+      rowRef: Ref,
+      keyFields: IndexedSeq[String]
+    ): ExtractionState = {
+      new ExtractionState(ctx, Env.empty[IR], rowRef, keyFields)
+    }
+  }
+
+  class ExtractionState(
+    val ctx: ExecuteContext,
+    env: Env[IR],
+    rowRef: Ref,
+    keyFields: IndexedSeq[String]
+  ) {
     val rowType: TStruct = rowRef.typ.asInstanceOf[TStruct]
     val rowKeyType: TStruct = rowType.select(keyFields)._1
     val firstKeyType: Type = rowKeyType.types.head
-    val iOrd: IntervalEndpointOrdering = rowKeyType.ordering(ctx.stateManager).intervalEndpointOrdering
+    val iOrd: IntervalEndpointOrdering = PartitionBoundOrdering(ctx, rowKeyType).intervalEndpointOrdering
 
     def getReferenceGenome(rg: String): ReferenceGenome = ctx.stateManager.referenceGenomes(rg)
 
-    def isFirstKey(ir: IR): Boolean = ir == GetField(rowRef, rowKeyType.fieldNames.head)
+    def bind(name: String, v: IR): ExtractionState =
+      new ExtractionState(ctx, env.bind(name, v), rowRef, keyFields)
+
+    private def isRowRef(ir: IR): Boolean = ir match {
+      case ref: Ref =>
+        ref == rowRef || env.lookupOption(ref.name).exists(isRowRef)
+      case _ => false
+    }
+
+    private def isKeyField(ir: IR, keyFieldName: String): Boolean = ir match {
+      case GetField(struct, name) => fieldIsKeyField(struct, name, keyFieldName)
+      case Ref(name, _) => env.lookupOption(name).exists(isKeyField(_, keyFieldName))
+      case _ => false
+    }
+
+    private def fieldIsKeyField(struct: IR, fieldName: String, keyFieldName: String): Boolean = struct match {
+      case MakeStruct(fields) => fields.exists { case (n, f) =>
+        n == fieldName && isKeyField(f, keyFieldName)
+      }
+      case SelectFields(o, fields) => fields.exists { f =>
+        f == fieldName && fieldIsKeyField(o, fieldName, keyFieldName)
+      }
+      case _ => isRowRef(struct) && fieldName == keyFieldName
+    }
+
+    def isFirstKey(ir: IR): Boolean = isKeyField(ir, rowKeyType.fieldNames.head)
 
     def isKeyStructPrefix(ir: IR): Boolean = ir match {
-      case MakeStruct(fields) => fields
-        .iterator
-        .map(_._2)
-        .zipWithIndex
-        .forall { case (fd, idx) => idx < rowKeyType.size && fd == GetField(rowRef, rowKeyType.fieldNames(idx)) }
-      case SelectFields(`rowRef`, fields) => keyFields.startsWith(fields)
+      case MakeStruct(fields) => fields.view.zipWithIndex.forall { case ((_, fd), idx) =>
```

Comment:
what's the `view` do?

Reply:
It makes a lazy collection wrapping `fields`, so the `zipWithIndex` doesn't materialize. See https://docs.scala-lang.org/overviews/collections/views.html